### PR TITLE
Update the test "cell-sort" to use Map

### LINKED ISSUE #704: Bag cell collections
The Haskell backend does not support Bag cell collections. Map cell collections are more commonly used, and are supported by the backend. We should either support bags, or modify the test to use maps.

### Instructions

```
$ git clone https://github.com/kframework/k

$ cd k; mvn package -DskipTests -Dllvm.backend.skip -Dhaskell.backend.skip

$ cd k-distribution/tests/regression-new

$ make KOMPILE_BACKEND=haskell --keep-going -C cell-sort
make: Entering directory '/home/ttuegel/k/k-distribution/tests/regression-new/cell-sort'
/home/ttuegel/k/k-distribution/bin/kompile  --backend haskell  test.k -d .
cat 1.test.in 2>/dev/null | /home/ttuegel/k/k-distribution/bin/krun 1.test   -d . | diff - 1.test.out
Reading the input file TimeSpec {sec = 0, nsec = 84648}
Parsing the file TimeSpec {sec = 0, nsec = 204892794}
Verifying the definition TimeSpec {sec = 0, nsec = 130573995}
Reading the input file TimeSpec {sec = 0, nsec = 65988}
Parsing the file TimeSpec {sec = 0, nsec = 184365}
Verifying the pattern TimeSpec {sec = 0, nsec = 232424}
kore-exec: While applying axiom:
    left:
        Lbl'-LT-'generatedTop'-GT-'{}(
            Lbl'-LT-'T'-GT-'{}(
                Lbl'-LT-'k'-GT-'{}(
                    kseq{}(
                        inj{SortStmt{}, SortKItem{}}(
                            LbladdContract'LParUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id{}(
                                VarContractName:SortId{}
                            )
                        ),
                        VarDotVar2:SortK{}
                    )
                ),
                VarDotVar1:SortFunctionsCell{},
                Lbl'-LT-'contracts'-GT-'{}(
                    Lbl'Unds'ContractCellBag'Unds'{}(
                        Lbl'Stop'ContractCellBag{}(),
                        VarDotVar3:SortContractCellBag{}
                    )
                )
            ),
            VarDotVar0:SortGeneratedCounterCell{}
        )
    right:
        Lbl'-LT-'generatedTop'-GT-'{}(
            Lbl'-LT-'T'-GT-'{}(
                Lbl'-LT-'k'-GT-'{}(VarDotVar2:SortK{}),
                VarDotVar1:SortFunctionsCell{},
                Lbl'-LT-'contracts'-GT-'{}(
                    Lbl'Unds'ContractCellBag'Unds'{}(
                        inj{SortContractCell{}, SortContractCellBag{}}(
                            Lbl'-LT-'contract'-GT-'{}(
                                Lbl'-LT-'cName'-GT-'{}(
                                    kseq{}(
                                        inj{SortId{}, SortKItem{}}(
                                            VarContractName:SortId{}
                                        ),
                                        dotk{}()
                                    )
                                ),
                                Lbl'-LT-'cFunctions'-GT-'{}(Lbl'Stop'Map{}())
                            )
                        ),
                        VarDotVar3:SortContractCellBag{}
                    )
                )
            ),
            VarDotVar0:SortGeneratedCounterCell{}
        )
    requires:
        \top{SortGeneratedTopCell{}}()
    ensures:
        \top{SortGeneratedTopCell{}}()
from the initial configuration:
    \and{_}(
        /* term: */
        Lbl'-LT-'generatedTop'-GT-'{}(
            Lbl'-LT-'T'-GT-'{}(
                Lbl'-LT-'k'-GT-'{}(
                    kseq{}(
                        inj{SortStmt{}, SortKItem{}}(
                            LbladdContract'LParUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id{}(
                                \dv{SortId{}}("foo")
                            )
                        ),
                        kseq{}(
                            inj{SortStmt{}, SortKItem{}}(
                                LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                    \dv{SortId{}}("foo"),
                                    \dv{SortId{}}("f1")
                                )
                            ),
                            kseq{}(
                                inj{SortStmt{}, SortKItem{}}(
                                    LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                        \dv{SortId{}}("foo"),
                                        \dv{SortId{}}("f2")
                                    )
                                ),
                                kseq{}(
                                    inj{SortStmt{}, SortKItem{}}(
                                        LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                            \dv{SortId{}}("foo"),
                                            \dv{SortId{}}("f3")
                                        )
                                    ),
                                    kseq{}(
                                        inj{SortStmt{}, SortKItem{}}(
                                            LbltestFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                                \dv{SortId{}}("foo"),
                                                \dv{SortId{}}("f2")
                                            )
                                        ),
                                        dotk{}()
                                    )
                                )
                            )
                        )
                    )
                ),
                Lbl'-LT-'functions'-GT-'{}(
                    Lbl'-LT-'nextFunction'-GT-'{}(\dv{SortInt{}}("0")),
                    Lbl'Stop'FunctionCellBag{}()
                ),
                Lbl'-LT-'contracts'-GT-'{}(Lbl'Stop'ContractCellBag{}())
            ),
            Lbl'-LT-'generatedCounter'-GT-'{}(\dv{SortInt{}}("0"))
        ),
        \and{_}(
            /* predicate: */
            \and{_PREDICATE{}}(
                \ceil{SortContractCellBag{}, _PREDICATE{}}(
                    Lbl'Stop'ContractCellBag{}()
                ),
                \ceil{SortFunctionCellBag{}, _PREDICATE{}}(
                    Lbl'Stop'FunctionCellBag{}()
                )
            ),
            /* substitution: */
            \top{_PREDICATE{}}()
        )
    )
with the unifier:
    \and{_}(
        /* term: */
        \top{_PREDICATE{}}(),
        \and{_}(
            /* predicate: */
            \and{_PREDICATE{}}(
                \and{_PREDICATE{}}(
                    \ceil{SortContractCellBag{}, _PREDICATE{}}(
                        Lbl'Unds'ContractCellBag'Unds'{}(
                            Lbl'Stop'ContractCellBag{}(),
                            VarDotVar3:SortContractCellBag{}
                        )
                    ),
                    \ceil{SortFunctionCellBag{}, _PREDICATE{}}(
                        Lbl'Stop'FunctionCellBag{}()
                    )
                ),
                \equals{SortContractCellBag{}, _PREDICATE{}}(
                    Lbl'Unds'ContractCellBag'Unds'{}(
                        Lbl'Stop'ContractCellBag{}(),
                        VarDotVar3:SortContractCellBag{}
                    ),
                    Lbl'Stop'ContractCellBag{}()
                )
            ),
            /* substitution: */
            \and{_PREDICATE{}}(
                \and{_PREDICATE{}}(
                    \and{_PREDICATE{}}(
                        \equals{SortId{}, _PREDICATE{}}(
                            VarContractName:SortId{},
                            \dv{SortId{}}("foo")
                        ),
                        \equals{SortGeneratedCounterCell{}, _PREDICATE{}}(
                            VarDotVar0:SortGeneratedCounterCell{},
                            Lbl'-LT-'generatedCounter'-GT-'{}(
                                \dv{SortInt{}}("0")
                            )
                        )
                    ),
                    \equals{SortFunctionsCell{}, _PREDICATE{}}(
                        VarDotVar1:SortFunctionsCell{},
                        Lbl'-LT-'functions'-GT-'{}(
                            Lbl'-LT-'nextFunction'-GT-'{}(\dv{SortInt{}}("0")),
                            Lbl'Stop'FunctionCellBag{}()
                        )
                    )
                ),
                \equals{SortK{}, _PREDICATE{}}(
                    VarDotVar2:SortK{},
                    kseq{}(
                        inj{SortStmt{}, SortKItem{}}(
                            LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                \dv{SortId{}}("foo"),
                                \dv{SortId{}}("f1")
                            )
                        ),
                        kseq{}(
                            inj{SortStmt{}, SortKItem{}}(
                                LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                    \dv{SortId{}}("foo"),
                                    \dv{SortId{}}("f2")
                                )
                            ),
                            kseq{}(
                                inj{SortStmt{}, SortKItem{}}(
                                    LbladdFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                        \dv{SortId{}}("foo"),
                                        \dv{SortId{}}("f3")
                                    )
                                ),
                                kseq{}(
                                    inj{SortStmt{}, SortKItem{}}(
                                        LbltestFunction'LParUndsCommUndsRParSClnUnds'TEST-SYNTAX'UndsUnds'Id'Unds'Id{}(
                                            \dv{SortId{}}("foo"),
                                            \dv{SortId{}}("f2")
                                        )
                                    ),
                                    dotk{}()
                                )
                            )
                        )
                    )
                )
            )
        )
    )
Failed substitution coverage check!
The substitution (above, in the unifier) did not cover the axiom variables:
    VarDotVar3:SortContractCellBag{}
in the left-hand side of the axiom.
CallStack (from HasCallStack):
  error, called at src/Kore/Step/Step.hs:498:6 in kore-0.0.1.0-JoeANYTIHtF5l52ggMmeRc:Kore.Step.Step
[Error] Critical: I/O Error while executing
(./.krun-2019-06-06-11-24-55-387-a7afb17a-97bd-4699-b570-aa92e6c91cd8/result.kore
(No such file or directory))
make: Leaving directory '/home/ttuegel/k/k-distribution/tests/regression-new/cell-sort'
```

### PR DESCRIPTION
**Pull request by andreiburdusa**
Fixes: kframework/kore#704

### CODE REVIEW

File: k-distribution/tests/regression-new/cell-sort/test.k [hunk 1]
```diff
@@ -19,12 +19,12 @@ module TEST
                   <k color="green"> $PGM:Pgm </k>
                   <functions color="blue">
                     <nextFunction> 0:Int </nextFunction> // It works again if we move this outside of <functions>
```
Comment (andreiburdusa) [inline k-distribution/tests/regression-new/cell-sort/test.k]: Should I move it outside of \<functions\> so the test would work?

Comment (ttuegel) [reply k-distribution/tests/regression-new/cell-sort/test.k]: Yes, I don't think this syntax is supported by the Kore backends. At least, I've never seen it used by the semantics that we try to support.

Comment (andreiburdusa) [reply k-distribution/tests/regression-new/cell-sort/test.k]: I think something changed. The test is now passing and in order to keep the changes I had to make before, I had to change `1.test.out`

**Comment by ttuegel** [comment]
@andreiburdusa Does this work when you `make KOMPILE_BACKEND=haskell`? If so, please make a `cell-sort-haskell` test directory with:

1. a symbolic link to `cell-sort/test.k` and
2. a `Makefile` that sets `KOMPILE_BACKEND = haskell`.

Look at [the Bytes test](https://github.com/kframework/k/tree/master/k-distribution/tests/regression-new/bytes-haskell) for an example.

**Comment by ehildenb** [comment]
Unless it's absolutely necessary, you shouldn't be updating the Haskell backend submodule in this PR. I think it's _unlikely_ that it's necessary, because the Haskell backend is tested against an _older_ commit of K, so the updates to K should for the most part be able to wait for the haskell-backend submodule update job instead of updating itself.

I'll re-iterate as well, _anyone_ can approve the submodule update jobs, it does _not_ have to be the person requested for review. So if you need the submodule update job to go through and @ttuegel is not around ot approve it, feel free to approve it yourself.

**Comment by ttuegel** [comment]
The submodule update is not necessary for this pull request.

**Comment by andreiburdusa** [comment]
@ehildenb @ttuegel 
Sorry, I don't know what action of mine caused the update. And I thought I reverted that using `git rebase -i origin/master`. Maybe I'll just close this pr and open a new one

**Comment by ehildenb** [comment]
@andreiburdusa another simple option is to just add another commit on top which corrects the submodule commit. Figure out the commit that `origin/master` of K uses for haskell backend, checkout that commit in the submodule, and then use `git add haskell-backend/src/main/native/haskell-backend` to add the submodule, then commit it. Then when you push you should see no update to the submodule here anymore.

**Comment by ehildenb** [comment]
BTW commit b61336b75b1deed8d7b139176177c8d39def9d78 updates the submodule.

Also, I would recommend _not_ using `git add -u` or `git commit -a`, because if you have any changes to the submodule they'll be added.

Instead, explicitely `git add PATH/TO/FILE` that you need every time, and then do `git commit -m "MESSAGE"`

**Comment by ehildenb** [comment]
It still appears that there is a change to the LLVM backend submodule included here as well, which should be un-done.

**Comment by andreiburdusa** [comment]
> @andreiburdusa Does this work when you `make KOMPILE_BACKEND=haskell`? If so, please make a `cell-sort-haskell` test directory with:
> 
>     1. a symbolic link to `cell-sort/test.k` and
> 
>     2. a `Makefile` that sets `KOMPILE_BACKEND = haskell`.
> 
> 
> Look at [the Bytes test](https://github.com/kframework/k/tree/master/k-distribution/tests/regression-new/bytes-haskell) for an example.

No, it doesn't. The Java backend gives the output
```
<T>
  <k>
    "done"
  </k>
  <nextFunction>
    3
  </nextFunction>
  <functions>
    <fId>
      0
    </fId> |-> <function>
      <fId>
        0
      </fId>
    </function>
    <fId>
      1
    </fId> |-> <function>
      <fId>
        1
      </fId>
    </function>
    <fId>
      2
    </fId> |-> <function>
      <fId>
        2
      </fId>
    </function>
  </functions>
  <contracts>
    <cName>
      foo
    </cName> |-> <contract>
      <cName>
        foo
      </cName>
      <cFunctions>
        f1 |-> 0
        f2 |-> 1
        f3 |-> 2
      </cFunctions>
    </contract>
  </contracts>
</T>
```
while the Haskell backend
```
<T>
  <k>
    "done"
  </k>
  <nextFunction>
    3
  </nextFunction>
  <functions>
    <function>
      <fId>
        0
      </fId>
    </function> <function>
      <fId>
        1
      </fId>
    </function> <function>
      <fId>
        2
      </fId>
    </function>
  </functions>
  <contracts>
    <contract>
      <cName>
        foo
      </cName>
      <cFunctions>
        f1 |-> 0
        f2 |-> 1
        f3 |-> 2
      </cFunctions>
    </contract>
  </contracts>
</T>
```

File: k-distribution/tests/regression-new/cell-sort/1.test.out [hunk 2]
```diff
@@ -6,32 +6,22 @@
     3
   </nextFunction>
   <functions>
-    <fId>
-      0
-    </fId> |-> <function>
+    <function>
```
Comment (ttuegel) [inline k-distribution/tests/regression-new/cell-sort/1.test.out]: This should be the correct output with the Java backend.

**Review by ttuegel — CHANGES_REQUESTED** [review]
We should not change the Java backend test at all.

**Review by ttuegel — APPROVED** [review]

**Pull request merged**